### PR TITLE
DaemonSetJob: implement jobTemplate update

### PR DESCRIPTION
**Pull request by clamoriniere1A**
Issue: #71

- Add JobSpec md5 hash in Job annotation.
- Add more unit-tests for daemonsetjobcontroller.Sync method
- Add end2end test for DaemonSetJob

### CODE REVIEW

File: hack/run-test-e2e.sh [hunk 1]
```diff
@@ -30,4 +30,4 @@ popd
 #  - check if kubeconfig is there
 #  - check if cluster is up
 
-./test/e2e/e2e.test --kubeconfig=$HOME/.kube/config
+./test/e2e/e2e.test --kubeconfig=$HOME/.kube/config --ginkgo.slowSpecThreshold 60
```
Comment (sdminonne) [inline hack/run-test-e2e.sh]: ouff... OK :(

File: pkg/controller/common.go [hunk 2]
```diff
@@ -29,3 +31,15 @@ func deleteJobsFromLabelSelector(namespace string, jobsSelector labels.Selector,
 	}
 	return utilerrors.NewAggregate(errs)
 }
+
+func compareJobSpecMD5Hash(hash string, job *batch.Job) bool {
```
Comment (sdminonne) [inline pkg/controller/common.go]: this deserve more love ;)

File: pkg/controller/daemonsetjob.go [hunk 3]
```diff
@@ -355,6 +380,7 @@ func (d *DaemonSetJobController) manageDaemonSetJob(daemonsetjob *dapi.DaemonSet
 }
 
 func getJobsStatus(jobs []*batch.Job) (activeJobs, succeededJobs, failedJobs int32) {
+
```
Comment (sdminonne) [inline pkg/controller/daemonsetjob.go]: ?

File: pkg/controller/daemonsetjob_test.go [hunk 4]
```diff
@@ -13,9 +13,13 @@ import (
 	dclientset "github.com/amadeusitgroup/workflow-controller/pkg/client/clientset/versioned"
 	winformers "github.com/amadeusitgroup/workflow-controller/pkg/client/informers/externalversions"
 	dlisters "github.com/amadeusitgroup/workflow-controller/pkg/client/listers/daemonsetjob/v1"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
```
Comment (sdminonne) [inline pkg/controller/daemonsetjob_test.go]: thanks!

File: pkg/controller/daemonsetjob.go [hunk 5]
```diff
@@ -288,6 +289,11 @@ func (d *DaemonSetJobController) manageDaemonSetJob(daemonsetjob *dapi.DaemonSet
 	daemonsetjobComplete := false
 	daemonsetjobFailed := false
 
+	md5JobSpec, err := generateMD5JobSpec(&daemonsetjob.Spec.JobTemplate.Spec)
+	if err != nil {
+		return err
```
Comment (sdminonne) [inline pkg/controller/daemonsetjob.go]: add error context: `fmt.Errorf("... %v")`

File: pkg/controller/job_control.go [hunk 6]
```diff
@@ -57,7 +62,19 @@ func (w WorkflowJobControl) CreateJobFromDaemonSetJob(namespace string, template
 	}
 	templateCopy := template.DeepCopy()
 	templateCopy.Spec.Template.Spec.NodeName = nodeName
-	job, err := w.CreateJob(namespace, nodeName, &daemonsetjob.ObjectMeta, templateCopy, &labels)
+
+	// Generate a MD5 representing the JobTemplateSpec send
+	hash, err := generateMD5JobSpec(&daemonsetjob.Spec.JobTemplate.Spec)
+	if err != nil {
+		return nil, err
```
Comment (sdminonne) [inline pkg/controller/job_control.go]: add error context. `fmt.Errorf("... %v", err)`

File: test/e2e/daemonsetjobs.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+})
+
+func deleteDaemonSetJob(workflowClient versioned.Interface, daemonsetjob *dapi.DaemonSetJob) {
+	workflowClient.DaemonsetjobV1().DaemonSetJobs(daemonsetjob.Namespace).Delete(daemonsetjob.Name, nil)
+	By("DaemonSetJob deleted")
+}
+
+func deleteAllJobsFromDaemonSetJob(kubeClient clientset.Interface, daemonsetjob *dapi.DaemonSetJob) {
+	jobs, err := kubeClient.Batch().Jobs(daemonsetjob.Namespace).List(metav1.ListOptions{
+		LabelSelector: controller.DaemonSetJobLabelKey + "=" + daemonsetjob.Name,
```
Comment (sdminonne) [inline test/e2e/daemonsetjobs.go]: just curious. We don't have a function that instantiate the proper label?

File: test/e2e/framework/daemonsetjobs_utils.go [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+		}
+		Logf("DaemonSetJob created")
+		return nil
+	}
+}
+
+// HOIsDaemonSetJobStarted is an higher order func that returns the func that checks whether DaemonSetJob is started
+func HOIsDaemonSetJobStarted(workflowClient versioned.Interface, daemonsetjob *daemonsetjobv1.DaemonSetJob) func() error {
+	return func() error {
+
```
Comment (sdminonne) [inline test/e2e/framework/daemonsetjobs_utils.go]: ?

File: hack/run-test-e2e.sh (hunk 1 above)
Comment (clamoriniere1A) [reply hack/run-test-e2e.sh]: 30s is maybe enough :)

File: pkg/controller/common.go (hunk 2 above)
Comment (clamoriniere1A) [reply pkg/controller/common.go]: which kind of love?

File: pkg/controller/daemonsetjob.go (hunk 5 above)
Comment (clamoriniere1A) [reply pkg/controller/daemonsetjob.go]: right

File: pkg/controller/job_control.go (hunk 6 above)
Comment (clamoriniere1A) [reply pkg/controller/job_control.go]: right

File: test/e2e/daemonsetjobs.go (hunk 7 above)
Comment (clamoriniere1A) [reply test/e2e/daemonsetjobs.go]: I should use some apimachinery struct, like labels.Set and then convert it to a String

File: pkg/controller/daemonsetjob.go (hunk 3 above)
Comment (clamoriniere1A) [reply pkg/controller/daemonsetjob.go]: right

File: pkg/controller/daemonsetjob_test.go (hunk 4 above)
Comment (clamoriniere1A) [reply pkg/controller/daemonsetjob_test.go]: you are welcome :)

File: test/e2e/framework/daemonsetjobs_utils.go (hunk 8 above)
Comment (clamoriniere1A) [reply test/e2e/framework/daemonsetjobs_utils.go]: done

**Comment by clamoriniere1A** [comment]
@sdminonne let me know if you have others remarks after my changes

**Comment by sdminonne** [comment]
thanks!

**Pull request merged**